### PR TITLE
MCP23XXX Refactor

### PR DESCRIPTION
**Pull request by jesserockz**
## Description:

This refactors the `MCP23008`, `MCP23017`, `MCP23S08` and `MCP23S17` to remove duplicate code and inherit from new base classes.

It also removes the duplicate pin schemas and consolidates all 4 into a single new schema with key `mcp23xxx` for the component id.

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1028

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by kbx81** [comment]
@jesserockz this is looking good! I want to test it out but time keeps getting away from me. I rolled through the code, though, and I like how you've (re)arranged things. 😄

File: esphome/components/mcp23016/__init__.py [hunk 1]
```diff
@@ -4,22 +4,28 @@
...
-}).extend(cv.COMPONENT_SCHEMA).extend(i2c.i2c_device_schema(0x20))
+CONFIG_SCHEMA = (
+    cv.Schema(
+        {
+            cv.Required(CONF_ID): cv.declare_id(MCP23016),
+        }
+    )
+    .extend(cv.COMPONENT_SCHEMA)
+    .extend(i2c.i2c_device_schema(0x20))
+)
```
Comment (glmnet) [inline esphome/components/mcp23016/__init__.py:28]: Are you modifying just the formatting on these? Why?

**Review by glmnet — CHANGES_REQUESTED** [review]
This pr has lots of changes with files with almost the same name and reviewing file changes with just formattings is quite annoying waste of time. I've seen other people already do this and asked him to rollback the format change, I'm ok with a formatter update on the codebase but that will be another PR

**Comment by jesserockz** [comment]
> This pr has lots of changes with files with almost the same name and reviewing file changes with just formattings is quite annoying waste of time. I've seen other people already do this and asked him to rollback the format change, I'm ok with a formatter update on the codebase but that will be another PR

Yeah sorry, I just ran `black mcp*`

File: esphome/components/mcp23016/__init__.py (hunk 1 above)
Comment (jesserockz) [reply esphome/components/mcp23016/__init__.py]: I reverted the mcp23016 file =)

**Pull request merged**